File: Java/palindrome-linked-list/Solution2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode prev = null;
        ListNode current = slow;
        ListNode next = null;

        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        ListNode first_half = head;
        ListNode second_half = prev;

        while (first_half != null && second_half != null) {
            if (first_half.val != second_half.val) {
                return false;
            }

            first_half = first_half.next;
            second_half = second_half.next;
        }

        return true;
    }
}

File: Python/palindrome-linked-list/solution2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        list = []
        current = head

        while current is not None:
            list.append(current.val)
            current = current.next

        left = 0
        right = len(list) - 1

        while left < right:
            if list[left] != list[right]:
                return False
            
            left += 1
            right -= 1

        return True